torch_glow/src/PyTorchCommon.cpp

```diff
@@ -188,7 +215,138 @@ at::Tensor glowTypeToEmptyPTTensor(const glow::Type &glowType) {
+    } else {
+      LOG(DFATAL)
+          << "PyTorch tensor with unsupported quantization scheme detected.";
+    }
+    if (ptTensor.scalar_type() == at::kQUInt8) {
+      offset -= OFFSETSHIFT;
+    }
+    auto glowType =
+        ptTypeToGlowType(*c10::TensorType::create(ptTensor), scale, offset);
+    return glow::Tensor(ptTensor.data_ptr(), &glowType);
+  } else {
+    auto glowType = ptTypeToGlowType(*c10::TensorType::create(ptTensor));
+    return glow::Tensor(ptTensor.data_ptr(), &glowType);
+  }
+}
+
+at::Tensor glowTensorToPTTensor(const glow::Tensor &glowTensor) {
+  std::vector<int64_t> sizes;
+  for (const auto dim : glowTensor.dims()) {
+    sizes.push_back(dim);
+  }
+  return at::from_blob(glowTensor.getUnsafePtr(), sizes,
+                       at::device(at::kCPU).dtype(at::kByte));
 }
+
+std::tuple<at::Tensor, c10::optional<at::Tensor>>
+fbgemm_linear_unpack(at::Tensor packed_weight) {
+
+  // TODO zrphercule:
+  // We should enbale the following code instead of using the pointer directly.
+  // auto& pack_ptr =
+  // at::cpp_custom_type_hack::cast<PackedLinearWeight>(packed_weight);
+  PackedLinearWeight &pack_ptr = *reinterpret_cast<PackedLinearWeight *>(
+      packed_weight.storage().data_ptr().get());
+  auto packB = pack_ptr.w.get();
+
+  int64_t N = static_cast<int64_t>(packB->numCols());
+  int64_t K = static_cast<int64_t>(packB->numRows());
+
+  at::Tensor weight_origin;
+  if (pack_ptr.q_scheme == at::kPerTensorAffine) {
+    weight_origin = at::_empty_affine_quantized(
+        {N, K}, at::device(at::kCPU).dtype(at::kQInt8), pack_ptr.w_scale[0],
+        pack_ptr.w_zp[0]);
+  } else if (pack_ptr.q_scheme == at::kPerChannelAffine) {
+    auto scales =
+        at::from_blob(pack_ptr.w_scale.data(), pack_ptr.w_scale.size(),
+                      at::device(at::kCPU).dtype(at::kFloat));
+    auto zero_points = at::from_blob(pack_ptr.w_zp.data(), pack_ptr.w_zp.size(),
+                                     at::device(at::kCPU).dtype(at::kInt));
+
+    weight_origin = at::_empty_per_channel_affine_quantized(
+        {N, K}, scales.toType(at::kDouble), zero_points.toType(at::kLong),
+        0, // The output channel axis is 0
+        at::device(at::kCPU).dtype(at::kQInt8));
+  }
+
+  int8_t *weight_ptr_int8 =
+      reinterpret_cast<int8_t *>(weight_origin.data_ptr<c10::qint8>());
+
+  packB->unpack(weight_ptr_int8);
+
+  return std::tuple<at::Tensor, c10::optional<at::Tensor>>(weight_origin,
+                                                           pack_ptr.bias);
+}
+
+std::tuple<at::Tensor, c10::optional<at::Tensor>>
+fbgemm_conv_unpack(at::Tensor packed_weights, const int kSpatialDim) {
+
+  // TODO zrphercule:
+  // We should enbale the following code instead of using the pointer directly.
+  // auto& pack_ptr =
+  // at::cpp_custom_type_hack::cast<PackedConvWeight<2>>(packed_weights);
+  PackedConvWeight<2> &pack_ptr = *reinterpret_cast<PackedConvWeight<2> *>(
```

Comment:
What's 2 here?

Reply:
2 indicate 2d, and 3 indicate 3d. Since we did enbaled 3d yet.

Reply:
Can you document this in the comments?

Reply:
Sure!